### ISSUE TITLE
Proposal: add support for aggregating 

### ISSUE THREAD

**Issue by tpaschalis** [grafana/agent#3804]
## Background 

The loki.process component allows for defining a [multiline stage](https://grafana.com/docs/agent/latest/flow/reference/components/loki.process/#stagemultiline-block) which enables to merge consecutive log lines into a single log entry in a processing pipeline.

While the concept is immediately useful in cases we want to collect closely related lines (eg. coming from a stack trace), it is quite limited in what it can do, since it can only match a certain starting regex and collapse consecutive entries until a `max_wait_time` or `max_lines` has elapsed, making it hard to use for network-based sources where entries might arrive interleaved or with some delay.

I propose that we add a new processing stage (or even a separate component) that can more flexibly aggregate log entries based on a set of conditions. Incoming log entries would _not_ be immediately forwarded to the next processing stage or output component, but would be buffered in memory or on a temporary file on disk, allowing more entries to be collected and finally pushed as an aggregated entry.

The stage's name is up for grabs, but I'm thinking something like `aggregate`, `mix`, `bundle`, `accumulate`, or `collect`

## Blueprint
* This new stage starts by defining and keeping track of a specified `map_key` from the extracted set of values.
* If the `map_key` does not exist, then the stage is a no-op for the specified log entry, and is forwarded to the next processing stage immediately.
* If the `map_key` exists, the entry is buffered under the key's value and stored separately. This allows interleaved entries to be correctly matched together. A conditional expression could signal what we want to do with any incoming log entries if the `map_key` already exists; either append to the buffered, or overwriting it and creating a new one.
* Incoming entries that match the `map_key` are appended the entry separated by the newline, and their extracted map of values is merged into the currently existing extracted map.
* The stage configures a condition (most likely based on another extracted value) that will signify the end of the aggregation, where the set of entries will then be propagated to the next stage or component. As a fallback the stage will still feature the `max_wait_time` and `max_lines` options.
* After the log entry is forwarded, it is evicted from the temporary storage (either in-memory or a temporary disk file).

## Use cases
* Aggregating closely-related log lines to mitigate noisy logging output
* Aggregating entries from a Kafka topic that belong to the same job
* Extract HTTP headers throughout a user-journey ID and present them together
* Recording timestamps belonging to a certain user-journey ID to produce the total journey duration

## Example

Let's consider the following two stages 
```
loki.json {
  expressions = {tx = "transaction_id", tx_end = "transaction_status"}
}
loki.aggregate {
  map_key = "tx"
  end_label = "tx_end"
  end_value = "done"
}
```

and these incoming log lines.

```
```

## Thoughts
* This could also be a separate component in itself, but I feel it
* Something similar could also be applied to metrics, once #2783 is in place. This would allow us to aggregate or filter metrics at the Agent level, which is an idea that has come up in the past as well (#834).
* If we enable clustering for logging components, we'll need to ensure that every event from a specific stream ends up to the same Agent so that the aggregation is not cut short due to a cluster change.
